File: HashMap.js
// DEFINE HASH MAP CLASS
class HashMap {
  constructor(initialCapacity = 16, loadFactor = 0.75) {
    this.buckets = new Array(initialCapacity);
    this.size = 0;
    this.loadFactor = loadFactor;
  }

  // DEFINE HASH FUNCTION
  hash(key) {
    let hashCode = 0;
    const primeNumber = 31;
    for (let i = 0; i < key.length; i++) {
      hashCode = (primeNumber * hashCode + key.charCodeAt(i)) % this.buckets.length;
    }
    return hashCode;
  }

  // SET KEY-VALUE PAIR
  set(key, value) {
    const index = this.hash(key);
    if (index < 0 || index >= this.buckets.length) {
      throw new Error("Trying to access index out of bound");
    }

    if (!this.buckets[index]) {
      this.buckets[index] = [];
    }

    for (let i = 0; i < this.buckets[index].length; i++) {
      if (this.buckets[index][i][0] === key) {
        this.buckets[index][i][1] = value;
        return;
      }
    }

    this.buckets[index].push([key, value]);
    this.size++;

    if (this.size > this.buckets.length * this.loadFactor) {
      this.grow();
    }
  }

  // GET VALUE BY KEY
  get(key) {
    const index = this.hash(key);
    if (this.buckets[index]) {
      for (let i = 0; i < this.buckets[index].length; i++) {
        if (this.buckets[index][i][0] === key) {
          return this.buckets[index][i][1];
        }
      }
    }
    return null;
  }

  // CHECK IF KEY EXISTS
  has(key) {
    const index = this.hash(key);
    if (this.buckets[index]) {
      for (let i = 0; i < this.buckets[index].length; i++) {
        if (this.buckets[index][i][0] === key) {
          return true;
        }
      }
    }
    return false;
  }

  // REMOVE KEY-VALUE PAIR
  remove(key) {
    const index = this.hash(key);
    if (this.buckets[index]) {
      for (let i = 0; i < this.buckets[index].length; i++) {
        if (this.buckets[index][i][0] === key) {
          this.buckets[index].splice(i, 1);
          this.size--;
          return true;
        }
      }
    }
    return false;
  }

  // GET LENGTH OF HASH MAP
  length() {
    return this.size;
  }

  // CLEAR HASH MAP
  clear() {
    this.buckets = new Array(this.buckets.length);
    this.size = 0;
  }

  // GET KEYS
  keys() {
    const keysArray = [];
    for (let i = 0; i < this.buckets.length; i++) {
      if (this.buckets[i]) {
        for (let j = 0; j < this.buckets[i].length; j++) {
          keysArray.push(this.buckets[i][j][0]);
        }
      }
    }
    return keysArray;
  }

  // GET VALUES
  values() {
    const valuesArray = [];
    for (let i = 0; i < this.buckets.length; i++) {
      if (this.buckets[i]) {
        for (let j = 0; j < this.buckets[i].length; j++) {
          valuesArray.push(this.buckets[i][j][1]);
        }
      }
    }
    return valuesArray;
  }

  // GET ENTRIES
  entries() {
    const entriesArray = [];
    for (let i = 0; i < this.buckets.length; i++) {
      if (this.buckets[i]) {
        for (let j = 0; j < this.buckets[i].length; j++) {
          entriesArray.push([this.buckets[i][j][0], this.buckets[i][j][1]]);
        }
      }
    }
    return entriesArray;
  }

  // RESIZE BUCKETS WHEN LOAD FACTOR IS REACHED
  grow() {
    const newCapacity = this.buckets.length * 2;
    const newBuckets = new Array(newCapacity);
    for (let i = 0; i < this.buckets.length; i++) {
      if (this.buckets[i]) {
        for (let j = 0; j < this.buckets[i].length; j++) {
          const [key, value] = this.buckets[i][j];
          const newIndex = this.hash(key);
          if (!newBuckets[newIndex]) {
            newBuckets[newIndex] = [];
          }
          newBuckets[newIndex].push([key, value]);
        }
      }
    }
    this.buckets = newBuckets;
  }
}

// CREATE HASH MAP INSTANCE
const map = new HashMap();
map.set("key1", "value1");
map.set("key2", "value2");
console.log(map.get("key1")); // Output: value1
console.log(map.has("key2")); // Output: true
map.remove("key2");
console.log(map.keys()); // Output: ['key1']
console.log(map.values()); // Output: ['value1']
console.log(map.entries()); // Output: [['key1', 'value1']]
console.log(map.length()); // Output: 1
map.clear();
console.log(map.length()); // Output: 0

File: hashmap.rb
class HashMap
  def initialize
    @buckets = Array.new(16)
    @size = 0
    @load_factor = 0.75
  end

  # DEFINE HASH FUNCTION
  def hash(key)
    hash_code = 0
    prime_number = 31

    key.each_char { |char| hash_code = prime_number * hash_code + char.ord }

    hash_code
  end

  # SET KEY-VALUE PAIR
  def set(key, value)
    index = hash(key) % @buckets.length
    @buckets[index] ||= []

    @buckets[index].each_with_index do |pair, i|
      if pair[0] == key
        @buckets[index][i][1] = value
        return
      end
    end

    @buckets[index] << [key, value]
    @size += 1
    resize if @size >= @load_factor * @buckets.length
  end

  # GET VALUE BY KEY
  def get(key)
    index = hash(key) % @buckets.length
    return nil unless @buckets[index]

    @buckets[index].each do |pair|
      return pair[1] if pair[0] == key
    end
    nil
  end

  # CHECK IF KEY EXISTS
  def has(key)
    index = hash(key) % @buckets.length
    return false unless @buckets[index]

    @buckets[index].each do |pair|
      return true if pair[0] == key
    end
    false
  end

  # REMOVE KEY-VALUE PAIR
  def remove(key)
    index = hash(key) % @buckets.length
    return nil unless @buckets[index]

    @buckets[index].each_with_index do |pair, i|
      if pair[0] == key
        @size -= 1
        return @buckets[index].delete_at(i)[1]
      end
    end
    nil
  end

  # GET LENGTH OF HASH MAP
  def length
    @size
  end

  # CLEAR HASH MAP
  def clear
    @buckets = Array.new(16)
    @size = 0
  end

  # GET KEYS
  def keys
    result = []
    @buckets.each do |bucket|
      next unless bucket

      bucket.each do |pair|
        result << pair[0]
      end
    end
    result
  end

  # GET VALUES
  def values
    result = []
    @buckets.each do |bucket|
      next unless bucket

      bucket.each do |pair|
        result << pair[1]
      end
    end
    result
  end

  # GET ENTRIES
  def entries
    result = []
    @buckets.each do |bucket|
      next unless bucket

      bucket.each do |pair|
        result << pair
      end
    end
    result
  end

  private

  # RESIZE BUCKETS WHEN LOAD FACTOR IS REACHED
  def resize
    new_buckets = Array.new(@buckets.length * 2)
    @buckets.each do |bucket|
      next unless bucket

      bucket.each do |pair|
        index = hash(pair[0]) % new_buckets.length
        new_buckets[index] ||= []
        new_buckets[index] << pair
      end
    end
    @buckets = new_buckets
  end
end

# CREATE HASH MAP INSTANCE
hash_map = HashMap.new

# SET KEY-VALUE PAIR
hash_map.set("key1", "value1")
puts hash_map.entries.inspect

# GET VALUE BY KEY
puts hash_map.get("key1")

# CHECK IF KEY EXISTS
puts hash_map.has("key1")

# REMOVE KEY-VALUE PAIR
puts hash_map.remove("key1")

# GET LENGTH OF HASH MAP
puts hash_map.length

# CLEAR HASH MAP
hash_map.clear

# GET KEYS
puts hash_map.keys.inspect

# GET VALUES
puts hash_map.values.inspect

# GET ENTRIES
puts hash_map.entries.inspect
